static/src/javascripts/projects/common/modules/atoms/story-questions.js

```diff
@@ -0,0 +1,408 @@
+    const isEmailSubmissionReadyElement: ?HTMLElement = document.getElementById(
+        'js-storyquestion-is-email-submission-ready'
+    );
+    const isDeliveryTestReadyElement: ?HTMLElement = document.getElementById(
+        'js-storyquestion-is-answer-delivery-test-ready'
+    );
+
+    const isEmailSubmissionReady: boolean = !!(
+        isEmailSubmissionReadyElement &&
+        isEmailSubmissionReadyElement.dataset.isEmailSubmissionReady === 'true'
+    );
+    const isDeliveryTestReady: boolean = !!(
+        isDeliveryTestReadyElement &&
+        isDeliveryTestReadyElement.dataset.isAnswerDeliveryTestReady === 'true'
+    );
+
+    const askQuestionLinks: Element[] = [
+        ...document.querySelectorAll('.js-ask-question-link'),
+    ];
+
+    const answersDeliveryPreferences = document.querySelectorAll(
+        `.btn-answer-delivery-${atomId}`
+    );
+    const answerDeliveryPrefContainer = document.getElementById(
+        `js-delivery-selection-body-${atomId}`
+    );
+
+    askQuestionLinks.forEach(el => {
+        bean.on(el, 'click', event => {
+            askQuestion(event, isEmailSubmissionReady, isDeliveryTestReady);
+            el.classList.add('is-clicked');
+        });
+    });
+
+    if (answerDeliveryPrefContainer) {
+        const deliveryPrefList = [...answersDeliveryPreferences];
+
+        const pool = [0, 1, 2];
+        const flush = [];
+        while (pool.length > 0) {
+            const rand = Math.floor(Math.random() * pool.length);
+            flush.push(pool[rand]);
+            pool.splice(rand, 1);
+        }
+
+        flush.forEach(num => {
+            const relevantBtn = deliveryPrefList[num];
+            answerDeliveryPrefContainer.insertBefore(relevantBtn, null);
+        });
+
+        bean.one(
+            answerDeliveryPrefContainer,
+            'click',
+            deliveryPrefList,
+            event => {
+                if (!(event.target instanceof HTMLElement)) return;
+                submitDeliveryPreference(event);
+                (event.target: any).classList.add('is-clicked');
+            }
+        );
+    }
+
+    const finalCloseBtn: ?Element = document.getElementById(
+        `js-final-thankyou-message-${atomId}`
+    );
+
+    if (finalCloseBtn) {
+        bean.one(finalCloseBtn, 'click', event => {
+            event.preventDefault();
+            const storyQuestionAtom = document.getElementById(
+                `user__question-atom-${atomId}`
+            );
+            if (storyQuestionAtom) {
+                storyQuestionAtom.classList.add('is-hidden');
+            }
+            finalCloseBtn.classList.add('is-clicked');
+        });
+    }
+
+    [
+        ...document.querySelectorAll('.js-storyquestion-email-signup-form'),
+    ].forEach(el => {
+        bean.on(el, 'submit', submitSignUpForm);
+    });
+
+    getUserFromApi(userFromId => {
+        if (userFromId && userFromId.primaryEmailAddress) {
+            fastdom.write(() => {
+                [
```

Comment:
This shoud be wrapped in `fastdom.read()`

Reply:
Why that?

Reply:
Because

```js
...document.querySelectorAll('.js-storyquestion-email-signup-form')
```

reads from the DOM and doesn't write anything. This could be a good chance, to include `fastdom-promise` instead of `fastdom`, which we are trying to get rid of ...

```js
fastdom.read(() => document.querySelectorAll('.js-storyquestion-email-signup-form'))
.then(signup => fastdom.write(() => {
   /* all the rest of the noodle */
}))

Reply:
arf, this is a common misconception I've seen lying around a few times, and it needs correcting. Reading the DOM is absolutely fine here, because it does not trigger layout. The point of `read`/`write` is that fastdom makes sure operations that trigger layout all happen before any mutation of the same properties. This is because a mutation will invalidate previous calculation, and the browser will be forced to do another round of calculation if there is a read after it. So:
```js
el.clientHeight; // this is fine
el.appendChild(...); // changes geometry
el.clientHeight; // boom: forces relayout
```
is bad, that's what fastdom was written to prevent. Fortunately querying the DOM does not case any trouble, esp. `querySelectorAll` as it returns a _non-live_ collection. That would be a major hit for rendering speed if we had to do that. You can see the few properties that need to be sealed in `read` [here](https://gist.github.com/paulirish/5d52fb081b3570c81e3a).

I'm happy to move the query out of the `write` if you think it is necessary.

Reply:
tis true afaik – getting a nodes (or list of them) doesn't trigger layout, so this should be fine